File: src/main/java/com/allaboutjava/chap09/Tutorial_03_Chain_Futures_FlatMap.java
package com.allaboutjava.chap09;

import io.vavr.concurrent.Future;

import java.util.function.Function;

import static io.vavr.API.*;
import static io.vavr.Patterns.$Failure;
import static io.vavr.Patterns.$Success;
import static io.vavr.Predicates.instanceOf;
import static java.lang.Boolean.TRUE;
import static java.util.concurrent.TimeUnit.SECONDS;

public class Tutorial_03_Chain_Futures_FlatMap {

    public static void main(String[] args) throws InterruptedException {

        println("Step 1: Define a method which returns a Future");
        Function<String, Future<Integer>> donutStock = __ -> Future.of(() -> {
            Thread.sleep(500); // assume some long running database operation
            println("checking donut stock");
            return 10;
        });

        println("\nStep 2: Define another method which returns a Future");
        Function<Integer, Future<Boolean>> buyDonuts = quantity -> Future.of(() -> {
            // assume some long running database operation
            Thread.sleep(500);
            printf("buying %s donuts", quantity);
            return TRUE;
        });

        println("\nStep 3: Chaining Futures using flatMap");
        final Future<Boolean> buyingDonuts = donutStock.apply("vanilla donut").flatMap(buyDonuts::apply);
        Boolean isSuccess = buyingDonuts.await(5, SECONDS).get();
        printf("\nBuying vanilla donut was successful = %s", isSuccess);
        Thread.sleep(3000);
    }
}

File: src/main/scala/com/allaboutscala/chap09/Tutorial_03_Chain_Futures_FlatMap.scala
package com.allaboutscala.chap09

import scala.concurrent.Await
import scala.concurrent.duration._
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.language.postfixOps

/**
 * Chain futures using flatMap
 */
object Tutorial_03_Chain_Futures_FlatMap extends App {

  println("Step 1: Define a method which returns a Future")
  def donutStock(donut: String): Future[Int] = Future {
    // assume some long running database operation
    println("checking donut stock")
    10
  }

  println("\nStep 2: Define another method which returns a Future")
  def buyDonuts(quantity: Int): Future[Boolean] = Future {
    println(s"buying $quantity donuts")
    true
  }

  println("\nStep 3: Chaining Futures using flatMap")
  // To sequence multiple futures in order, you can make use of the flatMap() method. In our
  // example, we will chain the two future operations donutStock() and buyDonuts() by using
  // flatMap() method as shown below.
  val buyingDonuts: Future[Boolean] = donutStock("plain donut")
                                            .flatMap(qty => buyDonuts(qty))
  val isSuccess = Await.result(buyingDonuts, 5 seconds)
  println(s"Buying vanilla donut was successful = $isSuccess")
}
